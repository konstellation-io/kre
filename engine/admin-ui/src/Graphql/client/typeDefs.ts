import { AccessLevel, LogLevel } from 'Graphql/types/globalTypes';

export interface NodeSelection {
  workflowName: string;
  nodeNames: string[];
  __typename: string;
}

export interface LogPanelFilters {
  dateOption?: string;
  startDate?: string;
  endDate?: string;
  nodes?: NodeSelection[];
  search?: string;
  levels?: LogLevel[] | null;
  versionsIds?: string[] | null;
  workflowsNames?: string[] | null;
  __typename: 'logTabFilters';
}

export interface LogPanel {
  runtimeId: string;
  runtimeName: string;
  versionId: string;
  versionName: string;
  uniqueId: string;
  filters: LogPanelFilters;
}

export interface UserSettingsFilters {
  email: string | null;
  accessLevel: AccessLevel | null;
  __typename: 'UserSettingsFilters';
}
