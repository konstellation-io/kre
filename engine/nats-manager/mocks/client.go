// Code generated by MockGen. DO NOT EDIT.
// Source: client.go

// Package mocks is a generated GoMock package.
package mocks
