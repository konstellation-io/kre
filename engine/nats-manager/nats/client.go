package nats

import (
	"fmt"
	"regexp"

	"github.com/nats-io/nats.go"

	"github.com/konstellation-io/kre/engine/nats-manager/internal/entity"
	"github.com/konstellation-io/kre/engine/nats-manager/internal/errors"
	"github.com/konstellation-io/kre/engine/nats-manager/internal/logging"
)

type NatsClient struct {
	js     nats.JetStreamContext
	logger logging.Logger
}

func New(logger logging.Logger, js nats.JetStreamContext) *NatsClient {
	return &NatsClient{
		logger: logger,
		js:     js,
	}
}

func InitJetStreamConnection(url string) (nats.JetStreamContext, error) {
	natsConn, err := nats.Connect(url)
	if err != nil {
		return nil, fmt.Errorf("error connecting to NATS: %w", err)
	}

	js, err := natsConn.JetStream()
	if err != nil {
		return nil, fmt.Errorf("error connecting to NATS JetStream: %w", err)
	}

	return js, nil
}

func (n *NatsClient) CreateStream(streamConfig *entity.StreamConfig) error {
	n.logger.Infof("Creating stream  %q", streamConfig.Stream)

	subjects := n.getNodesSubjects(streamConfig.Nodes)

	streamCfg := &nats.StreamConfig{
		Name:        streamConfig.Stream,
		Description: "",
		Subjects:    append(subjects, streamConfig.EntrypointSubject),
		Retention:   nats.InterestPolicy,
	}

	_, err := n.js.AddStream(streamCfg)
	return err
}

func (n *NatsClient) CreateObjectStores(objectStore string) error {
	n.logger.Infof("Creating object store %q", objectStore)

	_, err := n.js.CreateObjectStore(&nats.ObjectStoreConfig{
		Bucket:  objectStore,
		Storage: nats.FileStorage,
	})
	if err != nil {
		return fmt.Errorf("error creating the object store: %w", err)
	}

	return nil
}

// GetObjectStoreNames returns the list of object stores.
// The optional param `optFilter` accepts 0 or 1 value.
func (n *NatsClient) GetObjectStoreNames(optFilter ...*regexp.Regexp) ([]string, error) {
	if len(optFilter) > 1 {
		return nil, errors.ErrNoOptFilter
	}

	var regexpFilter *regexp.Regexp
	if len(optFilter) == 1 {
		regexpFilter = optFilter[0]
	}

	objectStoresCh := n.js.ObjectStores()
	objectStores := make([]string, 0)

	for objectStore := range objectStoresCh {
		objStoreName := objectStore.Bucket()

		nameMatchFilter := regexpFilter == nil || regexpFilter.MatchString(objStoreName)
		if nameMatchFilter {
			objectStores = append(objectStores, objStoreName)
		}
	}

	return objectStores, nil
}

func (n *NatsClient) CreateObjectStore(objectStore string) error {
	n.logger.Infof("Creating object store %q", objectStore)

	_, err := n.js.CreateObjectStore(&nats.ObjectStoreConfig{
		Bucket:  objectStore,
		Storage: nats.FileStorage,
	})
	if err != nil {
		return fmt.Errorf("error creating the object store: %w", err)
	}

	return nil
}

func (n *NatsClient) CreateKeyValueStore(keyValueStore string) error {
	n.logger.Infof("Creating key-value store %q", keyValueStore)

	_, err := n.js.CreateKeyValue(&nats.KeyValueConfig{
		Bucket: keyValueStore,
	})
	if err != nil {
		return fmt.Errorf("error creating the key-value store: %s", err)
	}

	return nil
}

func (n *NatsClient) DeleteStream(stream string) error {
	n.logger.Infof("Deleting stream %q", stream)
	err := n.js.DeleteStream(stream)
	return err
}

func (n *NatsClient) DeleteObjectStore(objectStore string) error {
	n.logger.Infof("Deleting object store %q", objectStore)
	err := n.js.DeleteObjectStore(objectStore)
	return err
}

// GetStreamNames returns the list of streams' names.
// The optional param `optFilter` accepts 0 or 1 value.
func (n *NatsClient) GetStreamNames(optFilter ...*regexp.Regexp) ([]string, error) {
	if len(optFilter) > 1 {
		return nil, errors.ErrNoOptFilter
	}

	var regexpFilter *regexp.Regexp
	if len(optFilter) == 1 {
		regexpFilter = optFilter[0]
	}

	streamsChannel := n.js.StreamNames()
	streams := make([]string, 0)

	for streamName := range streamsChannel {
		nameMatchFilter := regexpFilter == nil || regexpFilter.MatchString(streamName)
		if nameMatchFilter {
			streams = append(streams, streamName)
		}
	}

	return streams, nil
}

func (n *NatsClient) getNodesSubjects(nodes entity.NodesStreamConfig) []string {
	subjects := make([]string, 0, len(nodes)*2)

	for _, nodeCfg := range nodes {
		subSubject := nodeCfg.Subject + ".*"
		subjects = append(subjects, nodeCfg.Subject, subSubject)
	}

	return subjects
}
