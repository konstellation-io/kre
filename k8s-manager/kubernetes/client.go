package kubernetes

import (
	"log"
	"os"
	"path/filepath"

	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"

	"github.com/konstellation-io/kre/k8s-manager/config"
)

func NewClientset(cfg *config.Config) *kubernetes.Clientset {
	kubeConfig := newKubernetesConfig(cfg)

	// create the clientset
	clientset, err := kubernetes.NewForConfig(kubeConfig)
	if err != nil {
		log.Fatalf("Fatal error kubernetes config: %s", err)
	}

	return clientset
}

func NewDynamicClient(cfg *config.Config) dynamic.Interface {
	kubeConfig := newKubernetesConfig(cfg)

	dynClient, err := dynamic.NewForConfig(kubeConfig)
	if err != nil {
		log.Fatalf("Fatal error kubernetes config: %s", err)
	}

	return dynClient
}

func newKubernetesConfig(config *config.Config) *rest.Config {
	if config.Kubernetes.IsInsideCluster {
		log.Printf("Creating K8s config in-cluster")

		kubeConfig, err := rest.InClusterConfig()
		if err != nil {
			log.Fatalf("fatal error kubernetes config: %s", err)
		}

		return kubeConfig
	}

	log.Printf("Creating K8s config from local .kube/config")

	// NOTE: It works only with the default user's config, not even the exported KUBECONFIG value
	kubeConfigPath := filepath.Join(os.Getenv("HOME"), ".kube", "config")

	// use the current context in kubeConfigPath
	kubeConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
	if err != nil {
		log.Fatalf("fatal error kubernetes config: %s", err)
	}

	return kubeConfig
}
