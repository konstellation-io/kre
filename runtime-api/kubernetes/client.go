package kubernetes

import (
	"os"
	"path/filepath"

	"github.com/konstellation-io/kre/libs/simplelogger"

	"github.com/konstellation-io/kre/runtime-api/config"

	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)

type K8sManager struct {
	cfg    *config.Config
	logger *simplelogger.SimpleLogger
}

func NewK8sManager(cfg *config.Config, logger *simplelogger.SimpleLogger) *K8sManager {
	return &K8sManager{
		cfg,
		logger,
	}
}

func (k *K8sManager) NewClientset() *kubernetes.Clientset {
	kubeConfig := k.newKubernetesConfig()

	// create the clientset
	clientset, err := kubernetes.NewForConfig(kubeConfig)
	if err != nil {
		k.logger.Errorf("Fatal error kubernetes config: %s", err)
		os.Exit(1)
	}

	return clientset
}

func (k *K8sManager) newKubernetesConfig() *rest.Config {
	if k.cfg.Kubernetes.IsInsideCluster == true {
		k.logger.Info("Creating K8s config in-cluster")

		kubeConfig, err := rest.InClusterConfig()
		if err != nil {
			k.logger.Errorf("fatal error kubernetes config: %s", err)
			os.Exit(1)
		}

		return kubeConfig
	}

	k.logger.Info("Creating K8s config from local .kube/config")

	// NOTE: It works only with the default user's config, not even the exported KUBECONFIG value
	kubeConfigPath := filepath.Join(os.Getenv("HOME"), ".kube", "config")

	// use the current context in kubeConfigPath
	kubeConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
	if err != nil {
		k.logger.Errorf("fatal error kubernetes config: %s", err)
		os.Exit(1)
	}

	return kubeConfig
}
