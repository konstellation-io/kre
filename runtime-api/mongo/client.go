package mongo

import (
	"context"
	"os"
	"time"

	"github.com/konstellation-io/kre/libs/simplelogger"

	"github.com/konstellation-io/kre/runtime-api/config"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type MongoDB struct {
	cfg    *config.Config
	logger *simplelogger.SimpleLogger
	client *mongo.Client
}

func NewDB(cfg *config.Config, logger *simplelogger.SimpleLogger) *MongoDB {
	return &MongoDB{
		cfg,
		logger,
		nil,
	}
}

func (m *MongoDB) NewClient() *mongo.Client {
	client, err := mongo.NewClient(options.Client().ApplyURI(m.cfg.MongoDB.Address))
	if err != nil {
		m.logger.Error(err.Error())
		os.Exit(1)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	err = client.Connect(ctx)
	if err != nil {
		m.logger.Error(err.Error())
		os.Exit(1)
	}

	m.logger.Info("The MongoDB client was created")
	m.client = client

	return client
}

func (m *MongoDB) Disconnect() {
	m.logger.Info("MongoDB disconnecting...")

	if m.client == nil {
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
	defer cancel()

	err := m.client.Disconnect(ctx)
	if err != nil {
		m.logger.Errorf("Error disconnecting MongoDB: %s", err)
		return
	}

	m.logger.Info("Connection to MongoDB closed.")
}
