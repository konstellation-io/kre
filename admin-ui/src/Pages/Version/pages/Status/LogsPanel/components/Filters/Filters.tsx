import GroupSelect, {
  GroupSelectData
} from 'Components/Form/GroupSelect/GroupSelect';
import MultiSelect, {
  MultiSelectOption,
  SelectTheme
} from 'Components/Form/MultiSelect/MultiSelect';

import DateFilter from './components/DatesFilter/DateFilter';
import { GetLogTabs_logTabs_filters } from 'Graphql/client/queries/getLogs.graphql';
import Left from 'Components/Layout/Left/Left';
import LevelIcon from 'Components/LevelIcon/LevelIcon';
import { LogLevel } from 'Graphql/types/globalTypes';
import { NodeSelection } from 'Graphql/client/typeDefs';
import React from 'react';
import Right from 'Components/Layout/Right/Right';
import SearchSelect from 'Components/Form/SearchSelect/SearchSelect';
import cx from 'classnames';
import styles from './Filters.module.scss';
import useWorkflowsAndNodes from 'Hooks/useWorkflowsAndNodes';

// TODO: use GroupSelectData in the filters instead of [string]
function nodesSelectionToDoubleSelector(
  selections: NodeSelection[]
): GroupSelectData {
  const data: { [key: string]: string[] } = {};

  selections.forEach(({ workflowName, nodeNames }) => {
    data[workflowName] = nodeNames;
  });

  return data;
}

function doubleSelectorToNodesSelection(
  selections: GroupSelectData
): NodeSelection[] {
  const data: NodeSelection[] = Object.entries(selections).map(
    ([workflowName, nodeNames]) => ({
      workflowName,
      nodeNames,
      __typename: 'NodeSelection'
    })
  );

  return data;
}

const levelsOrdered = [
  LogLevel.INFO,
  LogLevel.DEBUG,
  LogLevel.WARN,
  LogLevel.ERROR
];

type Props = {
  versionId: string;
  updateFilters: Function;
  filterValues: GetLogTabs_logTabs_filters;
};
function Filters({ updateFilters, filterValues, versionId }: Props) {
  const { workflowsAndNodesNames } = useWorkflowsAndNodes(versionId);

  const levelOptions = levelsOrdered.map(
    (level: LogLevel) =>
      ({
        label: level,
        Icon: <LevelIcon level={level} />
      } as MultiSelectOption<LogLevel>)
  );

  function onNodeSelection(newSelection: GroupSelectData) {
    updateFilters({
      nodes: doubleSelectorToNodesSelection(newSelection)
    });
  }

  function onLevelSelection(newLevels: LogLevel[]) {
    updateFilters({ levels: newLevels });
  }

  function onSearchUpdate(newSearch: string) {
    updateFilters({ search: newSearch });
  }

  return (
    <div className={styles.container}>
      <Left className={cx(styles.filterContainer, styles.leftContainer)}>
        <div className={styles.searchFilter}>
          <SearchSelect
            options={[]}
            onChange={onSearchUpdate}
            value={filterValues.search}
            className={styles.searchForm}
            placeholder="Search"
            hideError
            hideLabel
            showSearchIcon
          />
        </div>
        <div className={styles.selectProcesses}>
          <GroupSelect
            options={workflowsAndNodesNames}
            formSelectedOptions={nodesSelectionToDoubleSelector(
              filterValues.nodes || []
            )}
            onChange={onNodeSelection}
            label=""
            placeholder="Select Processes"
            className={styles.selectProcessForm}
            hideError
            hideSelections
          />
        </div>
      </Left>
      <Right className={cx(styles.filterContainer, styles.rightFilters)}>
        <div className={styles.filterLevel}>
          <MultiSelect<LogLevel>
            onChange={onLevelSelection}
            label=""
            hideError
            placeholder="ALL LEVELS"
            selectionUnit="LEVEL"
            selectAllText="ALL LEVELS"
            options={levelOptions}
            formSelectedOptions={filterValues.levels || []}
            theme={SelectTheme.LIGHT}
          />
        </div>
        <DateFilter
          selectedOption={filterValues.dateOption}
          updateFilters={updateFilters}
          formStartDate={filterValues.startDate}
          formEndDate={filterValues.endDate}
        />
      </Right>
    </div>
  );
}

export default Filters;
