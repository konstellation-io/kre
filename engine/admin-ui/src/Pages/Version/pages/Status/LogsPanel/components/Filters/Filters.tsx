import {
  GroupSelect,
  GroupSelectData,
  Left,
  MultiSelect,
  MultiSelectOption,
  MultiSelectTheme,
  Right,
  SearchSelect
} from 'kwc';

import DateFilter from './components/DatesFilter/DateFilter';
import LevelIcon from 'Components/LevelIcon/LevelIcon';
import { LogLevel } from 'Graphql/types/globalTypes';
import { LogPanelFilters, NodeSelection } from 'Graphql/client/typeDefs';
import React from 'react';
import cx from 'classnames';
import styles from './Filters.module.scss';
import useWorkflowsAndNodes from 'Hooks/useWorkflowsAndNodes';

// TODO: use GroupSelectData in the filters instead of [string]
function nodesSelectionToDoubleSelector(
  selections: NodeSelection[]
): GroupSelectData {
  const data: { [key: string]: string[] } = {};

  selections.forEach(({ workflowName, nodeNames }) => {
    data[workflowName] = nodeNames;
  });

  return data;
}

function doubleSelectorToNodesSelection(
  selections: GroupSelectData
): NodeSelection[] {
  const data: NodeSelection[] = Object.entries(selections).map(
    ([workflowName, nodeNames]) => ({
      workflowName,
      nodeNames,
      __typename: 'NodeSelection'
    })
  );

  return data;
}

const levelsOrdered = [
  LogLevel.INFO,
  LogLevel.DEBUG,
  LogLevel.WARN,
  LogLevel.ERROR
];

type Props = {
  versionName: string;
  updateFilters: Function;
  filterValues: LogPanelFilters;
  runtimeId: string;
};
function Filters({ updateFilters, filterValues, versionName, runtimeId }: Props) {
  const { workflowsAndNodesNames } = useWorkflowsAndNodes(versionName, runtimeId);

  const levelOptions = levelsOrdered.map(
    (level: LogLevel) =>
      ({
        label: level,
        Icon: <LevelIcon level={level} />
      } as MultiSelectOption<LogLevel>)
  );

  function onNodeSelection(newSelection: GroupSelectData) {
    updateFilters({
      nodes: doubleSelectorToNodesSelection(newSelection)
    });
  }

  function onLevelSelection(newLevels: LogLevel[]) {
    updateFilters({ levels: newLevels });
  }

  function onSearchUpdate(newSearch: string) {
    updateFilters({ search: newSearch });
  }

  return (
    <div className={styles.container}>
      <Left className={cx(styles.filterContainer, styles.leftContainer)}>
        <div className={styles.searchFilter}>
          <SearchSelect
            options={[]}
            onChange={onSearchUpdate}
            value={filterValues.search}
            className={styles.searchForm}
            placeholder="Search"
            hideError
            hideLabel
            showSearchIcon
          />
        </div>
        <div className={styles.selectProcesses}>
          <GroupSelect
            options={workflowsAndNodesNames}
            formSelectedOptions={nodesSelectionToDoubleSelector(
              filterValues.nodes || []
            )}
            onChange={onNodeSelection}
            label=""
            placeholder="Select Processes"
            className={styles.selectProcessForm}
            hideError
            hideSelections
          />
        </div>
      </Left>
      <Right className={cx(styles.filterContainer, styles.rightFilters)}>
        <div className={styles.filterLevel}>
          <MultiSelect<LogLevel>
            onChange={onLevelSelection}
            label=""
            hideError
            placeholder="ALL LEVELS"
            selectionUnit="LEVEL"
            selectAllText="ALL LEVELS"
            options={levelOptions}
            formSelectedOptions={filterValues.levels || []}
            theme={MultiSelectTheme.LIGHT}
          />
        </div>
        <DateFilter
          selectedOption={filterValues.dateOption}
          updateFilters={updateFilters}
          formStartDate={filterValues.startDate}
          formEndDate={filterValues.endDate}
        />
      </Right>
    </div>
  );
}

export default Filters;
