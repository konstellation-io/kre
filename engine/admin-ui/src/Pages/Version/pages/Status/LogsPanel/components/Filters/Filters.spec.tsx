import React from 'react';
import Filters from './Filters';
import { shallow } from 'enzyme';
import * as apolloClient from '@apollo/client';

apolloClient.useQuery = () => ({});

describe('Filters', () => {
  let wrapper: any;

  const onDateChangeMock = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <Filters
        updateFilters={onDateChangeMock}
        versionName={'version'}
        filterValues={{
          __typename: 'logTabFilters',
          startDate: '2020-01-01',
          endDate: '2020-01-01',
          dateOption: '2020-01-01'
        }}
      />
    );
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
