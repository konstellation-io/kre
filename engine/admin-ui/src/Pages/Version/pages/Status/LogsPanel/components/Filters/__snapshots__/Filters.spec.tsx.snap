// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters matches snapshot 1`] = `
<div
  className="container"
>
  <ot
    className="filterContainer leftContainer"
  >
    <div
      className="searchFilter"
    >
      <Component
        className="searchForm"
        hideError={true}
        hideLabel={true}
        onChange={[Function]}
        options={Array []}
        placeholder="Search"
        showSearchIcon={true}
      />
    </div>
    <div
      className="selectProcesses"
    >
      <Component
        className="selectProcessForm"
        formSelectedOptions={Object {}}
        hideError={true}
        hideSelections={true}
        label=""
        onChange={[Function]}
        options={
          Object {
            "": Array [
              "entrypoint",
            ],
          }
        }
        placeholder="Select Processes"
      />
    </div>
  </ot>
  <ht
    className="filterContainer rightFilters"
  >
    <div
      className="filterLevel"
    >
      <Component
        formSelectedOptions={Array []}
        hideError={true}
        label=""
        onChange={[Function]}
        options={
          Array [
            Object {
              "Icon": <LevelIcon
                level="INFO"
              />,
              "label": "INFO",
            },
            Object {
              "Icon": <LevelIcon
                level="DEBUG"
              />,
              "label": "DEBUG",
            },
            Object {
              "Icon": <LevelIcon
                level="WARN"
              />,
              "label": "WARN",
            },
            Object {
              "Icon": <LevelIcon
                level="ERROR"
              />,
              "label": "ERROR",
            },
          ]
        }
        placeholder="ALL LEVELS"
        selectAllText="ALL LEVELS"
        selectionUnit="LEVEL"
        theme="light"
      />
    </div>
    <DateFilter
      formEndDate="2020-01-01"
      formStartDate="2020-01-01"
      selectedOption="2020-01-01"
      updateFilters={[MockFunction]}
    />
  </ht>
</div>
`;
